### PR TITLE
First step towards incremental reduction of query responses

### PR DESCRIPTION
**Pull request by s1monw**
Today all query results are buffered up until we received responses of
all shards. This can hold on to a significant amount of memory if the number of
shards is large. This commit adds a first step towards incrementally reducing
aggregations results if a, per search request, configurable amount of responses
are received. If enough query results have been received and buffered all so-far
received aggregation responses will be reduced and released to be GCed.

this PR really needs some reviews and potential discussions but it's a start and outlines what it takes to make this feature work

### CODE REVIEW

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by s1monw** [comment]
@elasticmachine test this please

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 1]
```diff
@@ -593,4 +610,91 @@ public boolean isEmpty() {
...
+            return true;
+        }
+
+        private synchronized void consumeInternal(QuerySearchResult querySearchResult) {
+            assert hasAggs == null || hasAggs;
+            hasAggs = Boolean.TRUE;
+            if (buffer != null) {
+                InternalAggregations aggregations = (InternalAggregations) querySearchResult.consumeAggs();
+                // once the size is incremented to the length of the buffer we know all elements are added
+                // we also have happens before guarantees due to the memory barrier of the size write
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: is this comment outdated?

File: core/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java [hunk 2]
```diff
@@ -47,10 +47,21 @@
...
 
-        public ReduceContext(BigArrays bigArrays, ScriptService scriptService) {
+        public ReduceContext(BigArrays bigArrays, ScriptService scriptService, boolean isFinalReduce) {
             this.bigArrays = bigArrays;
             this.scriptService = scriptService;
+            this.isFinalReduce = isFinalReduce;
+        }
+
+        /**
+         * Returns <code>true</code> iff the current reduce phase is the final reduce phase. This indicated if operations like
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java]: s/indicated/indicates/

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/MinDocCountIT.java [hunk 3]
```diff
@@ -336,24 +336,8 @@ private void testMinDocCountOnTerms(String field, Script script, Terms.Order ord
...
-            try {
-                assertAllSuccessful(response);
-                assertSubset(allTerms, (Terms) response.getAggregations().get("terms"), minDocCount, size, include);
-            } catch (AssertionError ae) {
-                if (!retry) {
-                    throw ae;
-                }
-                logger.info("test failed. trying to see if it recovers after 1m.", ae);
-                try {
-                    Thread.sleep(60000);
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/MinDocCountIT.java]: !!!

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java [hunk 4]
```diff
@@ -228,7 +228,7 @@ public InternalAggregation doReduce(List<InternalAggregation> aggregations, Redu
             }
         }
 
-        final int size = Math.min(requiredSize, buckets.size());
+        final int size = reduceContext.isFinalReduce() == false ? buckets.size() : Math.min(requiredSize, buckets.size());
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java]: Since one of the goals of this change is to limit memory usage, I wonder whether this should use `getShardSize()` rather than `buckets.size()`, this should be a good trade-off between accuracy and memory usage? cc @colings86

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 5]
```diff
@@ -515,16 +525,23 @@ public final ReducedQueryPhase reducedQueryPhase(List<? extends AtomicArray.Entr
...
         final InternalAggregations aggregations = aggregationsList.isEmpty() ? null : reduceAggs(aggregationsList,
-            firstResult.pipelineAggregators());
+            firstResult.pipelineAggregators(), reduceContext);
         final SearchProfileShardResults shardResults = profileResults.isEmpty() ? null : new SearchProfileShardResults(profileResults);
         return new ReducedQueryPhase(totalHits, fetchHits, maxScore, timedOut, terminatedEarly, firstResult, suggest, aggregations,
             shardResults);
     }
 
+
+    private InternalAggregations reduceAggsOnly(List<InternalAggregations> aggregationsList) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: maybe update the name or add a comment to say that this method is about performing an intermediate reduce? (as opposed to final)

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 6]
```diff
@@ -459,12 +459,17 @@ private SearchHits getHits(ReducedQueryPhase reducedQueryPhase, boolean ignoreFr
...
+    }
+
     /**
      * Reduces the given query results and consumes all aggregations and profile results.
      * @see QuerySearchResult#consumeAggs()
      * @see QuerySearchResult#consumeProfileResult()
      */
-    public final ReducedQueryPhase reducedQueryPhase(List<? extends AtomicArray.Entry<? extends QuerySearchResultProvider>> queryResults) {
+    public final ReducedQueryPhase reducedQueryPhase(List<? extends AtomicArray.Entry<? extends QuerySearchResultProvider>> queryResults,
+                                                     List<InternalAggregations> reducedAggs) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: Can you add a comment to explain that reducedAggs is the result from intermediate reduce operations?

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 7]
```diff
@@ -475,9 +480,14 @@ public final ReducedQueryPhase reducedQueryPhase(List<? extends AtomicArray.Entr
...
-        final boolean hasAggs = firstResult.hasAggs();
+        final boolean hasAggs = firstResult.hasAggs() && reducedAggs == null;
         final boolean hasProfileResults = firstResult.hasProfileResults();
-        final List<InternalAggregations> aggregationsList = hasAggs ? new ArrayList<>(queryResults.size()) : Collections.emptyList();
+        final List<InternalAggregations> aggregationsList;
+        if (hasAggs) {
+            aggregationsList = new ArrayList<>(queryResults.size());
+        } else {
+            aggregationsList = reducedAggs == null ? Collections.emptyList() : reducedAggs;
+        }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: I think `hasAggs` is a confusing name given how this method was refactored. Maybe we should restructure the logic a bit with comments to explain what each case maps to? eg. something like

```java
if (reducedAggs != null) {
  // we already have results from intermediate reduces and just need to perform the final reduce
  assert firstResult.hasAggs();
} else if (firstResult.hasAggs()) {
  // the number of shards was less than the buffer size so we reduce agg results directly
} else {
  // no aggregations
}
```

**Review by jpountz — APPROVED** [review]
I like this change, I expected it to be more complex than that so this is a good surprise to me! I left some picky coments about naming and comments to make this change a bit easier to read. I think the interesting question is about how many buckets intermediate reduces for terms (or geo-hash) aggregations should produce.

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/MinDocCountIT.java (hunk 3 above)
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/MinDocCountIT.java]: dude!

**Comment by s1monw** [comment]
@elasticmachine would you bother to test this

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: yeah I had a complex solution first with non-blocking concurrency etc. I didn't go with it apparently

**Comment by s1monw** [comment]
@jpountz I pushed some changes

**Comment by jpountz** [comment]
Looks great. Do you have any opinions about the size to use for intermediate reduces of terms aggs? I'm good with pulling this change in and making it a follow-up, this change is already a net improvement as-is.

File: core/src/main/java/org/elasticsearch/action/search/SearchRequestBuilder.java [hunk 8]
```diff
@@ -46,6 +46,8 @@
  */
 public class SearchRequestBuilder extends ActionRequestBuilder<SearchRequest, SearchResponse, SearchRequestBuilder> {
 
+    private int reduceUpTo;
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/action/search/SearchRequestBuilder.java]: Is this used? It looks below like we set this directly on the request?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java (hunk 4 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java]: In theory I think this would be a good change to make, however I think we should do it in a separate PR as it may  require the error calculations to be tweaked a bit to be correct.

**Review by colings86 — APPROVED** [review]
LGTM, I left a couple of minor comments

**Comment by s1monw** [comment]
@jpountz I will open followup for the following things:

 * size of terms aggs in incremental reduce
 * exposing the batch size in REST (not part of this PR)
 * removing / raising our softlimit for num of shards in a search request
 * using a different collector if no aggs are requested
 * using the same technique for profiling results

I think we should just pull this one in without adding more stuff to it

**Comment by jpountz** [comment]
> I think we should just pull this one in without adding more stuff to it

yes

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by IdanWo** [comment]
Does this change effects the accuracy of the terms aggregation? Or it just technical takes a more fine-tuned approach in returning the results to the coordinating node?

**Comment by jpountz** [comment]
Yes, it potentially reduces the accuracy of `terms` aggregations.

Note that combined with https://github.com/elastic/elasticsearch/pull/25658, this change only starts reducing accuracy of the `terms` aggregations if more than 512 shards have matches. So say that you query 1000 shards but only 300 of them have matches, the accuracy will be the same as today.

**Comment by IdanWo** [comment]
Okay, got it. Will the user see this inaccuracy in the error bounds? (sum_other_doc_count, doc_count_error_upper_bound)?

I understand that there's nothing to do when making a request to so many shards at once, but I don't like the approach of "scaling out, performance and latency are much more important than accuracy". 
No one mention that less accurate results will come here, in Elasticsearch 5.4.0 released blog post:

> That said, it is quite easy to reach the 1,000 shard limit, especially with the recent release of Cross Cluster Search. As of 5.4.0, Top-N search results and aggregations are reduced in batches of 512, which puts an upper limit on the amount of memory used on the coordinating node, which has allowed us to set the shard soft limit to unlimited by default.

**Comment by jpountz** [comment]
> Will the user see this inaccuracy in the error bounds?

Yes.

> I understand that there's nothing to do when making a request to so many shards at once, but I don't like the approach of "scaling out, performance and latency are much more important factors than accuracy".

I would agree it would be discussable if that was only about performance and latency, but to me this is mostly about cluster stability, which I consider more important than the accuracy of `terms` aggs.

In my opinion the number of users that are affected by decreased accuracy of `terms` aggs is low enough that mentioning it in the release notes would be more confusing than helping.

**Comment by colings86** [comment]
@jpountz @IdanWo  Actually I am pretty sure that we don't lose any accuracy on the terms aggregation because of incremental reduce because we do not truncate the list of terms until we are doing the 'final' incremental reduction (see https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java#L284) so the accuracy should not be affected. We do have https://github.com/elastic/elasticsearch/issues/23285 open to explore truncating the list during the other incremental reductions though and this would indeed affect the accuracy of the terms aggregation if implemented.

**Comment by jpountz** [comment]
@colings86 Oh right, I remember we discussed it but had forgotten which approach we took. Thanks for clarifying!
